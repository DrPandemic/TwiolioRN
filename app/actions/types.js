// @flow

import { PhoneNumber, Message } from '../types';

export const SET_FETCHED_ACCOUNT_NUMBERS = 'SET_FETCHED_ACCOUNT_NUMBERS';
export const FETCH_ACCOUNT_NUMBERS = 'FETCH_ACCOUNT_NUMBERS';
export const FETCH_ACCOUNT_NUMBER_ERROR = 'FETCH_ACCOUNT_NUMBER_ERROR';

export const SELECT_NUMBER = 'SELECT_NUMBER';
export const SELECT_CONVERSATION = 'SELECT_CONVERSATION';

export const FETCH_MESSAGES = 'FETCH_MESSAGES';
export const SET_FETCHED_MESSAGES = 'SET_FETCHED_MESSAGES';
export const FETCH_MESSAGE_ERROR = 'FETCH_MESSAGE_ERROR';

export type FetchAccountNumbersT = {|
  type: 'FETCH_ACCOUNT_NUMBERS',
|}
export type SuccessFetchAccountNumbersT = {|
  type: 'SET_FETCHED_ACCOUNT_NUMBERS',
  fetchedAccountNumbers: Array<PhoneNumber>,
|}
export type FailFetchAccountNumbersT = {|
    type: 'FETCH_ACCOUNT_NUMBER_ERROR',
    error: any,
|}

export type SelectNumberT = {|
  type: 'SELECT_NUMBER',
  selectedNumber: ?string,
|}

export type FetchMessagesT = {|
  type: 'FETCH_MESSAGES',
|}
export type SuccessFetchMessagesT = {|
    type: 'SET_FETCHED_MESSAGES',
    fetchedMessages: Array<Message>,
|}
export type FailFetchMessagesT = {|
    type: 'FETCH_MESSAGE_ERROR',
    error: any,
|}

export type ActionT =
  | FetchAccountNumbersT
  | SuccessFetchAccountNumbersT
  | FailFetchAccountNumbersT
  | SelectNumberT
  | SelectConversationT
  | FetchMessagesT
  | SuccessFetchMessagesT
  | FailFetchMessagesT;
