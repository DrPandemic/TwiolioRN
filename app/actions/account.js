// @flow

import * as types from './types';
import Api from '../lib/api';

export function fetchAccountNumbers() {
  return (dispatch)  => Api.get('/IncomingPhoneNumbers.json')
    .then(r => r.json())
    .then(r => {
      dispatch({
        type: types.SET_FETCHED_ACCOUNT_NUMBERS,
        fetchedAccountNumbers : r.incoming_phone_numbers.map(number => new types.PhoneNumber(number))
      });
    })
    .catch(e => console.error('An error occured while fetching phone numbers', e));
}
