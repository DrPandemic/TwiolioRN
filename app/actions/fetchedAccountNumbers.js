// @flow

import * as types from './types';
import { PhoneNumber } from '../types';
import formatError from '../lib/errors';

export function fetchAccountNumbers(): types.FetchAccountNumbersT {
  return {
    type: types.FETCH_ACCOUNT_NUMBERS,
  };
}

export function successFetchAccountNumbers(numbers: Array<PhoneNumber>):
types.SuccessFetchAccountNumbersT {
  return {
    type: types.SET_FETCHED_ACCOUNT_NUMBERS,
    fetchedAccountNumbers: numbers,
  };
}

export function failFetchAccountNumbers(error: any):
types.FailFetchAccountNumbersT {
  return {
    type: types.FETCH_ACCOUNT_NUMBER_ERROR,
    error: formatError(error),
  };
}
