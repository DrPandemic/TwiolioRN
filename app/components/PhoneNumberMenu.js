// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Text } from 'react-native';
import {
  Menu,
  MenuOptions,
  MenuOption,
  MenuTrigger
} from 'react-native-popup-menu';

import { ActionCreators } from '../actions';
import type { T as FetchedT } from '../reducers/fetchedAccountNumbers';
import type { T as AccountT } from '../reducers/account';
import type { StateT } from '../reducers';
import { PhoneNumber } from '../types';

export class PhoneNumberMenu extends Component {
  props: {
    fetchedAccountNumbers: FetchedT,
    account: AccountT,
    selectNumber: ?string => void,
  }

  constructor(props: any) {
    super(props);

    this.onSelectNumber = this.onSelectNumber.bind(this);
  }

  onSelectNumber(number: ?string) {
    this.props.selectNumber(number);
  }

  renderNumbers() {
    if (this.props.fetchedAccountNumbers.numbers.length > 0) {
      return (
        <MenuOptions>
          <MenuOption value={null} key={'all'}>
            <Text>See all conversations</Text>
          </MenuOption>
          {
            this.props.fetchedAccountNumbers.numbers.map(
              n => PhoneNumberMenu.renderNumber(n)
            )
          }
        </MenuOptions>
      );
    }
    return (
      <MenuOptions>
        <MenuOption disabled={true} text={'You have no number'}>
        </MenuOption>
      </MenuOptions>
    );
  }

  static renderNumber(number: PhoneNumber) {
    return (
      <MenuOption value={number.sid} key={number.sid}>
        <Text>{number.friendlyName}</Text>
      </MenuOption>
    );
  }

  renderSelectedNumber() {
    const sid: ?string = this.props.account.selectedNumber;
    const defaultValue = 'All conversations';
    let displayName = defaultValue;

    if (sid !== null) {
      const foundNumber = this.props.fetchedAccountNumbers.numbers.find(
        n => n.sid === sid
      );

      if (foundNumber === undefined) {
        displayName = defaultValue;
      } else {
        displayName = foundNumber.friendlyName;
      }
    }

    return displayName;
  }

  render() {
    return (
      <Menu onSelect={this.onSelectNumber}>
        <MenuTrigger>
          <Text style={{ fontSize: 20 }}>
            {this.renderSelectedNumber()}...
          </Text>
        </MenuTrigger>
        {this.renderNumbers()}
      </Menu>
    );
  }
}

function mapStateToProps(state: StateT) {
  return {
    fetchedAccountNumbers: state.fetchedAccountNumbers,
    account: state.account,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(ActionCreators, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(PhoneNumberMenu);
