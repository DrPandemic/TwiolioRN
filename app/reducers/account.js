import createReducer from '../lib/createReducer'
import * as types from '../actions/types'

export const fetchAccountNumbers = createReducer({}, {
  [types.SET_ACCOUNT_NUMBER] (state, action) {
    return {
      ...state,
      accountNumbers: action.accountNumbers,
    };
  },
});

