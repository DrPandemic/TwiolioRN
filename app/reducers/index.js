import { combineReducers } from 'redux';
import * as AccountReducers from './account'

export default combineReducers(Object.assign({},
  AccountReducers,
));
