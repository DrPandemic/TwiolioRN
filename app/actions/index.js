import { push, goBack } from 'react-router-redux';

import * as fetchedAccountNumbersActions from './fetchedAccountNumbers';
import * as accountActions from './account';
import * as messageActions from './messages';
import * as persistActions from './persist';
import * as contactActions from './contacts';

export const ActionCreators = Object.assign(
  { push, goBack },
  fetchedAccountNumbersActions,
  accountActions,
  messageActions,
  persistActions,
  contactActions,
);
