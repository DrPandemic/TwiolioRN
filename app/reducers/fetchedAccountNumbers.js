// @flow

import { loop, Cmd } from 'redux-loop';
import * as types from '../actions/types';
import * as fetchActions from '../actions/fetchedAccountNumbers';
import { getApi } from '../lib/api';
import createReducer from '../lib/createReducer';
import effects from '../effects';
import { PhoneNumber } from '../types';

export type T = {|
  loading: boolean,
  numbers: Array<PhoneNumber>,
  error: any,
|};
export const initialState: T = {
  loading: false,
  numbers: [],
  error: {},
};

export const reducer = createReducer({
  [types.TICK](state: T) {
    return loop(
      { ...state },
      Cmd.action(fetchActions.fetchAccountNumbers()),
    );
  },
  [types.FETCH_ACCOUNT_NUMBERS](state: T) {
    return loop(
      { ...state, loading: true },
      Cmd.run(effects.fetchNumbers, {
        successActionCreator: fetchActions.successFetchAccountNumbers,
        failActionCreator: fetchActions.failFetchAccountNumbers,
        args: [getApi()],
      })
    );
  },
  [types.SET_FETCHED_ACCOUNT_NUMBERS](
    state: T,
    action: types.SuccessFetchAccountNumbersT
  ) {
    return {
      ...state,
      numbers: action.fetchedAccountNumbers,
      error: {},
      loading: false,
    };
  },
  [types.FETCH_ACCOUNT_NUMBER_ERROR](
    state: T,
    action: types.FailFetchAccountNumbersT
  ) {
    return {
      ...state,
      error: action.error,
      loading: false,
    };
  },
});
