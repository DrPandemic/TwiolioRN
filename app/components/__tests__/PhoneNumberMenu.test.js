import React from 'react';
import renderer from 'react-test-renderer';
import { Text } from 'react-native';

import { PhoneNumberMenu } from '../PhoneNumberMenu';
import phoneFixture from '../../test_helpers/fixtures/received_phone_number.json';
import { PhoneNumber } from '../../actions/types';
import { initialState as nInitial } from '../../reducers/fetchedAccountNumbers';
import { initialState as aInitial } from '../../reducers/account';

test('renders an empty menu', () => {
  const menu = new PhoneNumberMenu({
    fetchedAccountNumbers: { ...nInitial },
    account: { ...aInitial },
    selectNumber: () => {},
  });
  expect(renderer.create(menu.renderNumbers())).toMatchSnapshot();
});

test('renders menu with a number', () => {
  const menu = new PhoneNumberMenu({
    fetchedAccountNumbers: {
      ...nInitial,
      numbers: [new PhoneNumber(phoneFixture.simple)],
    },
    account: { ...aInitial },
    selectNumber: () => {},
  });
  expect(renderer.create(menu.renderNumbers())).toMatchSnapshot();
});

test('renders a selected number', () => {
  const menu = new PhoneNumberMenu({
    fetchedAccountNumbers: {
      ...nInitial,
      numbers: [new PhoneNumber(phoneFixture.simple)],
    },
    account: {
      ...aInitial,
      selectedNumber: phoneFixture.simple.sid,
    },
    selectNumber: () => {},
  });
  expect(renderer.create(
    <Text>
      {menu.renderSelectedNumber()}
    </Text>
  )).toMatchSnapshot();
});

test('renders an unselected number', () => {
  const menu = new PhoneNumberMenu({
    fetchedAccountNumbers: {
      ...nInitial,
      numbers: [new PhoneNumber(phoneFixture.simple)],
    },
    account: {
      ...aInitial,
      selectedNumber: null,
    },
    selectNumber: () => {},
  });
  expect(renderer.create(
    <Text>
      {menu.renderSelectedNumber()}
    </Text>
  )).toMatchSnapshot();
});
