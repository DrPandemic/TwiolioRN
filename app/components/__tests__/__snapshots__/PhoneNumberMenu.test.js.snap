// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a selected number 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  (123) 123-1234
</Text>
`;

exports[`renders an empty menu 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "padding": 5,
        },
        undefined,
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#ccc",
          },
          undefined,
        ]
      }
    >
      You have no number
    </Text>
  </View>
</View>
`;

exports[`renders an unselected number 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  All conversations
</Text>
`;

exports[`renders menu with a number 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "flex": 1,
            "padding": 5,
          },
          undefined,
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        See all conversations
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "flex": 1,
            "padding": 5,
          },
          undefined,
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        (123) 123-1234
      </Text>
    </View>
  </View>
</View>
`;
